README.md
# /r/weightroom latest daily redirect

## Required env vars

- CLIENT_ID: Reddit API client id
- CLIENT_SECRET: Reddit API secret


CHANGELOG.md
## [1.11.1](https://github.com/gish/wr-latest-daily-redirect/compare/v1.11.0...v1.11.1) (2022-07-20)


### Bug Fixes

* use correct argument for subreddit ([1d98894](https://github.com/gish/wr-latest-daily-redirect/commit/1d988940ede22e9c4a246b51726cfe75399b6267))

# [1.11.0](https://github.com/gish/wr-latest-daily-redirect/compare/v1.10.0...v1.11.0) (2022-07-20)


### Features

* add support for more subreddits ([4277737](https://github.com/gish/wr-latest-daily-redirect/commit/4277737d4c45c5b6546166a3396f2e29c052ca2f))

# [1.10.0](https://github.com/gish/wr-latest-daily-redirect/compare/v1.9.0...v1.10.0) (2022-07-20)


### Features

* respond with custom header with subreddit ([5151f3a](https://github.com/gish/wr-latest-daily-redirect/commit/5151f3a447fcd955ba829bb5f8caa328f0ed6be4))

# [1.9.0](https://github.com/gish/wr-latest-daily-redirect/compare/v1.8.0...v1.9.0) (2022-07-20)


### Features

* allow arbitrary subreddit ([b6a8d6f](https://github.com/gish/wr-latest-daily-redirect/commit/b6a8d6f9d1b9216ccc02a5f12500fa9b9be3daaa))

# [1.8.0](https://github.com/gish/wr-latest-daily-redirect/compare/v1.7.0...v1.8.0) (2022-07-20)


### Features

* remove cache ([7429537](https://github.com/gish/wr-latest-daily-redirect/commit/7429537a0f67996f7b2af307050b8e7037e26529))
* return generic error ([5d6bc9e](https://github.com/gish/wr-latest-daily-redirect/commit/5d6bc9ee9ef24646ea8b6c99c142e424ce50d270))
* support all subreddits ([e940c4a](https://github.com/gish/wr-latest-daily-redirect/commit/e940c4a1dd9046d20b452ce89271fef1f6db9f68))

# [1.7.0](https://github.com/gish/wr-latest-daily-redirect/compare/v1.6.1...v1.7.0) (2020-10-23)


### Features

* remove logging ([dbb6e4a](https://github.com/gish/wr-latest-daily-redirect/commit/dbb6e4a85c94e5a853619dcf34a2c4c39e80b4f1))

## [1.6.1](https://github.com/gish/wr-latest-daily-redirect/compare/v1.6.0...v1.6.1) (2020-02-29)


### Bug Fixes

* consider self posts only ([f2c037d](https://github.com/gish/wr-latest-daily-redirect/commit/f2c037dfc411fbc67b16e8131a16b439c3ac7f09))

# [1.6.0](https://github.com/gish/wr-latest-daily-redirect/compare/v1.5.1...v1.6.0) (2020-01-03)


### Features

* add bodybuilding sub ([0895dec](https://github.com/gish/wr-latest-daily-redirect/commit/0895dec46c2905128c6a9d379ea9dffb3cca5de2))

## [1.5.1](https://github.com/gish/wr-latest-daily-redirect/compare/v1.5.0...v1.5.1) (2019-12-13)


### Bug Fixes

* redirect to accessible origin ([2bcd8fa](https://github.com/gish/wr-latest-daily-redirect/commit/2bcd8fa83f1189259d3b2abf9612bd8ac47b1aa6))

# [1.5.0](https://github.com/gish/wr-latest-daily-redirect/compare/v1.4.0...v1.5.0) (2019-12-11)


### Features

* make brower forever cache old WR URL ([a300ba0](https://github.com/gish/wr-latest-daily-redirect/commit/a300ba07ca7d3ca665c07454580f08192e2b02f3))

# [1.4.0](https://github.com/gish/wr-latest-daily-redirect/compare/v1.3.1...v1.4.0) (2019-12-11)


### Features

* support fitness subreddit ([2744c08](https://github.com/gish/wr-latest-daily-redirect/commit/2744c081a9a433fb19fa09cf2514bd4bc02244bd))

## [1.3.1](https://github.com/gish/wr-latest-daily-redirect/compare/v1.3.0...v1.3.1) (2019-09-26)


### Bug Fixes

* always return latest cache object ([e9516b2](https://github.com/gish/wr-latest-daily-redirect/commit/e9516b2))

# [1.3.0](https://github.com/gish/wr-latest-daily-redirect/compare/v1.2.0...v1.3.0) (2019-09-24)


### Features

* cache redirect url ([660e083](https://github.com/gish/wr-latest-daily-redirect/commit/660e083))

# [1.2.0](https://github.com/gish/wr-latest-daily-redirect/compare/v1.1.3...v1.2.0) (2019-09-17)


### Features

* shorten response time ([21c70eb](https://github.com/gish/wr-latest-daily-redirect/commit/21c70eb))

## [1.1.3](https://github.com/gish/wr-latest-daily-redirect/compare/v1.1.2...v1.1.3) (2019-09-13)


### Bug Fixes

* push to git ([dc4eea3](https://github.com/gish/wr-latest-daily-redirect/commit/dc4eea3))

## [1.0.3](https://github.com/gish/wr-latest-daily-redirect/compare/v1.0.2...v1.0.3) (2019-09-12)


### Bug Fixes

* test ([17f6208](https://github.com/gish/wr-latest-daily-redirect/commit/17f6208))


src/redirect/getDaily/index.ts
import axios from "axios";
import qs from "qs";
import dotenv from "dotenv";
//import packagejson from "./../../../package.json";

dotenv.config();

const packagejson = { name: "wr-latest-daily-redirect", version: "1.7.0" };

const CLIENT_ID = process.env.CLIENT_ID ?? "";
const CLIENT_SECRET = process.env.CLIENT_SECRET ?? "";
const USER_AGENT = `nodejs:${packagejson.name}:${packagejson.version} (by /u/murrtu)`;

type RedditPost = {
  data: {
    is_self: boolean;
    title: string;
    url: string;
  };
};

const getAccessToken = async (username: string, password: string) => {
  const response = await axios({
    method: "POST",
    url: "https://www.reddit.com/api/v1/access_token",
    data: qs.stringify({ grant_type: "client_credentials" }),
    auth: {
      username,
      password,
    },
    headers: { "User-Agent": USER_AGENT },
  });
  return response.data.access_token;
};

const getNewPosts = (subreddit: string) => async (accessToken: string) => {
  const subredditPosts = await axios({
    method: "get",
    url: `https://oauth.reddit.com/r/${subreddit}/`,
    headers: {
      Authorization: "bearer " + accessToken,
      "User-Agent": USER_AGENT,
    },
  });
  return subredditPosts.data.data.children;
};

const findLatestDaily = (posts: RedditPost[]) => {
  const titleMatch = (post: RedditPost) => post.data.title.match(/daily/i);
  const selfMatch = (post: RedditPost) => post.data.is_self;
  const dailies = posts.filter((post) => selfMatch(post) && titleMatch(post));
  return dailies[0];
};

const getDailyUrl = async (
  subreddit: string,
  clientId: string,
  clientSecret: string
) => {
  const accessToken = await getAccessToken(clientId, clientSecret);
  const newPosts = await getNewPosts(subreddit)(accessToken);
  const latestDaily = await findLatestDaily(newPosts);
  const url = latestDaily.data.url;
  return url;
};

async function main(args: any) {
  try {
    const subreddit = args.subreddit ?? "weightroom";
    const url = await getDailyUrl(subreddit, CLIENT_ID, CLIENT_SECRET);
    return {
      headers: {
        location: url,
        "x-subreddit": subreddit,
      },
      statusCode: 302,
    };
  } catch (e) {
    console.error(e);
    if (axios.isAxiosError(e)) {
      return {
        statusCode: 501,
        body: e.message,
      };
    }
    return { statusCode: 500, body: e };
  }
}


index.js
const express = require('express');
const app = express();
const axios = require('axios');
const qs = require('qs');
const R = require('ramda');
const dotenv = require('dotenv');
const packagejson = require('./package.json');

dotenv.config();

const CLIENT_ID = process.env.CLIENT_ID;
const CLIENT_SECRET = process.env.CLIENT_SECRET;
const PORT = process.env.PORT;
const ORIGIN = process.env.ORIGIN;
const USER_AGENT = `nodejs:${packagejson.name}:${packagejson.version} (by /u/murrtu)`;
const PINTIFIER_KEY = process.env.PINTIFIER_KEY;
const CACHE_AGE = 23 * 60 * 60 * 1e3;
const SUBREDDITS = ['bodybuilding', 'fitness', 'weightroom'];

const createCache = () => {
  let storage = SUBREDDITS.reduce(
    (cache, subreddit) => ({
      ...cache,
      [subreddit]: {
        timestampUtc: 0,
        url: '',
      },
    }),
    {},
  );

  return {
    get: subreddit => storage[subreddit],
    update: subreddit => url =>
      (storage = {
        ...storage,
        [subreddit]: {
          timestampUtc: Date.now(),
          url,
        },
      }),
  };
};

const cache = createCache();

const getAccessToken = R.pipeP(
  (username, password) =>
    axios({
      method: 'POST',
      url: 'https://www.reddit.com/api/v1/access_token',
      data: qs.stringify({grant_type: 'client_credentials'}),
      auth: {
        username,
        password,
      },
      headers: {'User-Agent': USER_AGENT},
    }),
  R.path(['data', 'access_token']),
);

const getNewPosts = subreddit =>
  R.pipeP(
    accessToken =>
      axios({
        method: 'get',
        url: `https://oauth.reddit.com/r/${subreddit}/`,
        headers: {
          Authorization: 'bearer ' + accessToken,
          'User-Agent': USER_AGENT,
        },
      }),
    R.path(['data', 'data', 'children']),
  );

const findLatestDaily = posts => {
  const titleMatch = R.pipe(R.path(['data', 'title']), R.test(/daily/i));
  const selfMatch = R.pipe(R.path(['data', 'is_self']), R.equals(true));
  return R.pipe(R.filter(selfMatch), R.filter(titleMatch), R.head)(posts);
};

const addDailyToCache = subreddit => async (clientId, clientSecret) => {
  try {
    R.pipeP(
      await getAccessToken,
      await getNewPosts(subreddit),
      findLatestDaily,
      R.path(['data', 'url']),
      R.ifElse(
        R.equals(getDailyFromCache(cache.get())),
        R.identity,
        cache.update(subreddit),
      ),
    )(clientId, clientSecret);
  } catch (e) {
    console.error(e);
  }
};

const getDailyFromCache = R.path(['url']);
const getTimestampfromCache = R.path(['timestampUtc']);

const logVisit = (subreddit, dailyUrl, userAgent) =>
  axios({
    method: 'GET',
    url: 'https://pintifier.herokuapp.com/api/v1/notification',
    params: {
      key: PINTIFIER_KEY,
      domain: USER_AGENT,
      payload: JSON.stringify({subreddit, url: dailyUrl, ua: userAgent}),
    },
  });

SUBREDDITS.map(subreddit => {
  setInterval(
    () => addDailyToCache(subreddit)(CLIENT_ID, CLIENT_SECRET),
    1 * 60 * 1e3,
  );
  addDailyToCache(subreddit)(CLIENT_ID, CLIENT_SECRET);
});

app.get('/r/:subreddit', (req, res) => {
  try {
    const subreddit = req.params.subreddit;
    if (!SUBREDDITS.includes(subreddit)) {
      throw new Error(`Subreddit /r/${subreddit} not supported`);
    }
    const cachedSubreddit = cache.get(subreddit);
    const daily = getDailyFromCache(cachedSubreddit);
    const cacheTimestamp = getTimestampfromCache(cachedSubreddit);
    logVisit(subreddit, daily, req.get('User-Agent'));
    res
      .append(
        'Cache-Control',
        `max-age=${Math.floor(
          (cacheTimestamp + CACHE_AGE - Date.now()) / 1e3,
        )}`,
      )
      .status(302)
      .set('Location', daily)
      .end();
  } catch (e) {
    console.error(e);
    res
      .status(501)
      .send('Error: ' + e.message)
      .end();
  }
});

app.get('/', (_, res) => {
  res
    .status(301)
    .set('Location', `${ORIGIN}/r/weightroom`)
    .end();
});

app.listen(PORT);
